Connection established and closed callbacks

tewinget opened the pull request:
builds on #39

tewinget committed:
Add connection established and closed callbacks

Connection established is called when the connection has confirmed the
TLS handshake; server does this on ngtcp2 handshake "complete" after
sending the final packet for the handshake to the client which will then
call ngtcp2 handshake "confirmed".

Connection closed callback is called when a connection closes for any
reason, including that it timed out and was never actually established
(thus acting as a "connection attempt failed" signal as well).

In order to have a connection timeout call connection closed, a
handshake timeout needed to be set in ngtcp2 (the default was
funtionally infinite).  5s should be sufficient, but this can be made
configurable at a later time.

Tests which were confirming connection establishment or failure have
been updated to use the new callbacks.

dr7ana committed:
Connection open/close callback reworking

- no structs for open/close callbacks
- added `error.hpp`
- reworked bool_waiter in unit tests
- little bit of template magic from jason

dr7ana committed:
Added Endpoint::drop_connection

- Sometimes when ngtcp2 does not want a connection close packet written, we delete the connection from the endpoint's internal map
- In these cases, we still want to notify the calling endpoint that the connection was closed

jagerman commented on include/quic/error.hpp:
```diff
@@ -0,0 +1,82 @@
+
+        CONN_WRITE_CLOSE_FAIL = (1ULL << 62),
+        CONN_SEND_CLOSE_FAIL = (1ULL << 62) + 1
+    };
```
This needs a comment as to why 2^62 ± N is being used here as the underlying value.

dr7ana replied:
Resolved by off-band discussion, to-be fixed by Tom's next commits

tewinget committed:
work around incorrect ngtcp2 wakeup time

For whatever reason, sometimes ngtcp2 gives a time point in the past
when asked when to wake it up; this was resulting in us never waking it.
Normally this is fine, as an incoming packet wakes it and we do't notice,
but if no packets come in, new connections we make can fail to timeout
correctly.  This works around that issue by doing a 0µs timeout if the
timeout would be negative.

tewinget committed:
if connection is 'closing' on incoming packet, drop packet

we need to keep 'closing' connections around for a bit so we can drop
any straggler packets.  also, if we 'drop connection' here we get multiple
calls to the connection closed callback, which is not ideal

tewinget committed:
allow user-supplied connection close application error code

also use ngtcp2_ccerr_set_application_error rather then ...set_liberr if the
close is an application requested close rather than protocol error or similar

supply connection close error code to user-defined connection closed cb

adds connection close callback test

dr7ana commented on tests/009-close.cpp:
```diff
@@ -0,0 +1,52 @@
+        opt::local_addr server_local{};
+        opt::local_addr client_local{};
+
+        SECTION("Connection Closes with correct error code")
```
looks good to me! one question: are you planning on adding a second section to this test case?

tewinget replied:
nah, just couldn't remember if it had to have a section or not and figured it wouldn't hurt to put it regardless

dr7ana replied:
agreed

jagerman commented on src/connection.cpp:
```diff
@@ -234,21 +264,20 @@ namespace oxen::quic
+    void Connection::close_connection(uint64_t error_code)
     {
-        _endpoint.call([this]() { _endpoint.close_connection(*this); });
+        _endpoint.call([this, error_code]() { _endpoint.close_connection(*this, io_error{error_code}); });
```
Perhaps:
```suggestion
        _endpoint.call([this, err=io_error{error_code}]() { _endpoint.close_connection(*this, err); });
```

jagerman commented on src/connection.cpp:
```diff
@@ -757,19 +776,24 @@ namespace oxen::quic
         }
 
-        auto delta = exp_ns * 1ns - ts.time_since_epoch();
+        auto delta = std::chrono::nanoseconds(exp_ns * 1ns) - ts.time_since_epoch();
```
```suggestion
        auto delta = std::chrono::nanoseconds{exp_ns} - ts.time_since_epoch();
```

jagerman replied:
Or:
```suggestion
        auto delta = static_cast<int64_t>(exp_ns) * 1ns - ts.time_since_epoch();
```

jagerman commented on src/endpoint.cpp:
```diff
@@ -105,10 +117,24 @@ namespace oxen::quic
+
+        const auto* err = ngtcp2_conn_get_ccerr(conn);
+
+        std::string_view reason = "None"sv;
+        if (err->reasonlen > 0)
+            reason = std::string_view{reinterpret_cast<const char*>(err->reason), err->reasonlen};
```
```suggestion
        auto reason = err->reasonlen > 0
            ? std::string_view{reinterpret_cast<const char*>(err->reason), err->reasonlen}
            : "None"sv;
```

jagerman commented on src/endpoint.cpp:
```diff
@@ -105,10 +117,24 @@ namespace oxen::quic
+
+        const auto* err = ngtcp2_conn_get_ccerr(conn);
+
+        std::string_view reason = "None"sv;
+        if (err->reasonlen > 0)
+            reason = std::string_view{reinterpret_cast<const char*>(err->reason), err->reasonlen};
+
+        log::debug(log_cat, "Draining connection (CID: {}), Reason: {}", conn.scid(), reason);
```
```suggestion
        log::debug(log_cat, "Draining connection (CID: {}), Reason: {}", conn.scid(),
            err->reason ? std::string_view{reinterpret_cast<const char*>(err->reason), err->reasonlen} : "None"sv);
```

jagerman commented on src/endpoint.cpp:
```diff
@@ -154,39 +180,58 @@ namespace oxen::quic
+    {
+        const auto* err = ngtcp2_conn_get_ccerr(conn);
+
+        std::string_view reason = "None"sv;
+        if (err->reasonlen > 0)
+            reason = std::string_view{reinterpret_cast<const char*>(err->reason), err->reasonlen};
+
+        log::debug(log_cat, "Dropping connection (CID: {}), Reason: {}", conn.scid(), reason);
```
```suggestion
        log::debug(log_cat, "Dropping connection (CID: {}), Reason: {}", conn.scid(),
            err->reason ? std::string_view{reinterpret_cast<const char*>(err->reason), err->reasonlen} : "None"sv);
```

jagerman commented on src/endpoint.cpp:
```diff
@@ -198,8 +243,20 @@ namespace oxen::quic
+        ngtcp2_ccerr_default(&err);
+        if (ec.is_ngtcp2)
+            ngtcp2_ccerr_set_liberr(
+                    &err, ec.ngtcp2_code(), reinterpret_cast<uint8_t*>(const_cast<char*>(msg.data())), msg.size());
```
```suggestion
                    &err, ec.ngtcp2_code(), reinterpret_cast<const uint8_t*>(msg.data()), msg.size());
```

jagerman commented on src/endpoint.cpp:
```diff
@@ -198,8 +243,20 @@ namespace oxen::quic
+                    &err, ec.ngtcp2_code(), reinterpret_cast<uint8_t*>(const_cast<char*>(msg.data())), msg.size());
+        else
+            ngtcp2_ccerr_set_application_error(
+                    &err, ec.code(), reinterpret_cast<uint8_t*>(const_cast<char*>(msg.data())), msg.size());
```
```suggestion
                    &err, ec.code(), reinterpret_cast<const uint8_t*>(msg.data()), msg.size());
```

tewinget committed:
formatting and const correctness
